src/permutations.rs

```diff
@@ -0,0 +1,265 @@
+/// An iterator to iterate through all the `k`-permutations of a series of items.
+///
+/// Source items are distinguished by their position, not value; so if there
+/// are identical items in the source, there will be some identical permutation
+/// iterations.
+///
+/// Can be constructed from an in-memory list of items directly; or from an
+/// iterator, with the
+/// [`.permuatations()`](../trait.Itertools.html#method.permutations) method.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
```

Comment:
This annotation isn't strictly correct, given that:

> My `Permutations` implementations doesn't make use of `LazyBuffer`.

My only point of reluctance in merging this that the `Itertools::permutations` implementation collects `self`. Even without this annotation, there's a strong expectation that merely constructing an iterator adaptors are lazy and won't, for instance, loop infinitely (as this would on a cyclic iterator).

Nonetheless, I think this method is useful enough that it should be merged even in its non-lazy form, so long as:
 - this hazard is prominently documented
 - and there isn't any reason we couldn't make it lazy in the future without a breaking change.

Re. that second point: I'm worried making `PermutationSource` part of Itertool's public API exposes and commits us too much of `Permutations`'s internal implementation details.

So, I'd like to either:
1. merge a version `Permutations` that _is_ lazy, or
2. resolve my backward-compatibility concerns and merge a non-lazy `Permutations`.